piq/perceptual.py

```diff
@@ -0,0 +1,304 @@
+                and computing distance. See [2] for details.
+
+        References:
+            .. [1] Gatys, Leon and Ecker, Alexander and Bethge, Matthias
+            (2016). A Neural Algorithm of Artistic Style}
+            Association for Research in Vision and Ophthalmology (ARVO)
+            https://arxiv.org/abs/1508.06576
+    
+            .. [2] Zhang, Richard and Isola, Phillip and Efros, et al.
+            (2018) The Unreasonable Effectiveness of Deep Features as a Perceptual Metric
+            2018 IEEE/CVF Conference on Computer Vision and Pattern Recognition
+            https://arxiv.org/abs/1801.03924
+
+        """
+        super().__init__()
+
+        if callable(feature_extractor):
+            self.model = feature_extractor
+            self.layers = layers
+        else:
+            if feature_extractor == "vgg16":
+                self.model = vgg16(pretrained=True, progress=False).features
+                self.layers = [VGG16_LAYERS[l] for l in layers]
+            elif feature_extractor == "vgg19":
+                self.model = vgg19(pretrained=True, progress=False).features
+                self.layers = [VGG19_LAYERS[l] for l in layers]
+            else:
+                raise ValueError("Unknown feature extractor")
+
+        if use_average_pooling:
+            self.model = self.max_pool_to_average_pool(self.model)
+
+        # Disable gradients
+        for param in self.model.parameters():
+            param.requires_grad_(False)
+
+        self.distance = {
+            "mse": nn.MSELoss,
+            "mae": nn.L1Loss,
+        }[distance](reduction='none')
+
+        self.weights = [torch.tensor(w) for w in weights]
+        mean = torch.tensor(mean)
+        std = torch.tensor(std)
+        self.mean = mean.view(1, -1, 1, 1)
+        self.std = std.view(1, -1, 1, 1)
+        
+        self.normalize_features = normalize_features
+        self.reduction = reduction
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of Content loss between feature representations of prediction and target tensors.
+
+        Args:
+            prediction: Tensor of prediction of the network.
+            target: Reference tensor.
+
+        """
+        _validate_input(input_tensors=(prediction, target), allow_5d=False)
+        prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+        return self.compute_metric(prediction, target)
+    
+    def compute_metric(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        # Normalize input
+        mean, std = self.mean.to(prediction), self.std.to(prediction)
+        prediction = (prediction - mean) / std
+        target = (target - mean) / std
+
+        self.model.to(prediction)
+        prediction_features = self.get_features(prediction)
+        target_features = self.get_features(target)
```

Comment:
This code is duplicated -lines 230 - 236. 
Replace it with a new fuction/method or include this functionality in `self.get_features`.

Reply:
Unified interface for both classes by moving most of operations into `forward` call